src/table.js

```diff
@@ -538,15 +543,106 @@ export function getTypeValidator(colType) {
   }
 }
 
+export function coerceToType(value, type) {
+  const stringValue = typeof value === "string" ? value.trim() : value;
+  switch (type) {
+    case "raw":
+      return value;
+    case "string":
+      return typeof value === "string"
+        ? stringValue
+        : value || value === 0
+        ? value.toString()
+        : null;
+    case "boolean":
+      return value === true || stringValue === "true"
+        ? true
+        : value === false || stringValue === "false"
+        ? false
+        : null;
+    case "integer":
+      return value === 0
+        ? value
+        : !value || isNaN(parseInt(value))
+        ? NaN
+        : parseInt(value);
+    case "bigint":
+      return typeof value === "bigint"
+        ? value
+        : value === 0 || value === true || value === false
+        ? // eslint-disable-next-line no-undef
+          BigInt(value)
+        : !value || isNaN(value) || !Number.isInteger(+value)
+        ? NaN
+        : // eslint-disable-next-line no-undef
+          BigInt(value);
+    case "number": {
+      return value === 0
+        ? value
+        : !value || isNaN(value)
+        ? NaN
+        : Number(value);
+    }
+    case "date": {
+      if (value instanceof Date) return value;
+      if (typeof value === "string") {
+        let match;
+        if (
+          (match = stringValue.match(
+            /^([-+]\d{2})?\d{4}(-\d{2}(-\d{2})?)?(T\d{2}:\d{2}(:\d{2}(\.\d{3})?)?(Z|[-+]\d{2}:\d{2})?)?$/
+          ))
+        ) {
+          if (fixTz && !!match[4] && !match[7])
+            value = stringValue.replace(/-/g, "/").replace(/T/, " ");
+        }
+      }
+      return new Date(value);
+    }
+    case "array":
+      if (Array.isArray(value)) return value;
+      return [value];
+    case "object":
+      // this will return true for everything except null, undefined, strings,
+      // numbers, boolean, and symbols, so may yield unexpected results.
+      if (typeof value === "object") return value;
+      return {value: value};
+    case "buffer":
+    case "other":
+    default:
+      return value || value === 0 ? value : null;
+  }
+}
+
 // This function applies table cell operations to an in-memory table (array of
 // objects); it should be equivalent to the corresponding SQL query. TODO Use
 // DuckDBClient for data arrays, too, and then we wouldn’t need our own __table
 // function to do table operations on in-memory data?
 export function __table(source, operations) {
   const input = source;
   let {schema, columns} = source;
+  let newlyInferred = false;
```

Comment:
Maybe call this `inferredSchema`? Since we’ll re-infer this every time `__table` is called so there isn’t really a concept of “newly” here.